salt/crypt.py

```diff
@@ -1100,9 +1100,18 @@ def minion_sign_in_payload(self):
         payload["nonce"] = uuid.uuid4().hex
         if "autosign_grains" in self.opts:
             autosign_grains = {}
-            for grain in self.opts["autosign_grains"]:
-                autosign_grains[grain] = self.opts["grains"].get(grain, None)
-            payload["autosign_grains"] = autosign_grains
+        for grain in self.opts["autosign_grains"]:
+            value = self.opts["grains"].get(grain, None)
+            if isinstance(value, str):
+                autosign_grains[grain] = value
+            elif isinstance(value, (int, float, bool)):
+                autosign_grains[grain] = str(value)
+            elif isinstance(value, dict):
+                autosign_grains[grain] = dict.__str__(value)
+        else:
+            autosign_grains[grain] = str(value)
```

Comment:
Is the indentation off here? Or do you want this to be a `for-else` clause? In that case both `grain` and `value` can be undefined if `self.opts["autosign_grains"]` is empty